sale_stock_coupon/models/sale_order.py

```diff
@@ -0,0 +1,227 @@
+        elif program.discount_apply_on == "on_order":
+            self._set_delivered_reward_qty_for_program_discount_percentage_on_order(
+                program
+            )
+
+    def _set_delivered_reward_qty_for_program_discount(self, program):
+        if program.discount_type == "fixed_amount":
+            self._set_delivered_reward_qty_for_program_discount_fixed_amount(program)
+        elif program.discount_type == "percentage":
+            self._set_delivered_reward_qty_for_program_discount_percentage(program)
+
+    def _set_delivered_reward_qty_for_program(self, program):
+        if program.reward_type == "product":
+            self._set_delivered_reward_qty_for_program_product(program)
+        elif program.reward_type == "discount":
+            self._set_delivered_reward_qty_for_program_discount(program)
+
+    def recompute_coupon_lines(self):
+        super().recompute_coupon_lines()
+
+        applied_programs = self._get_applied_programs_with_rewards_on_current_order()
+        order_lines = (
+            self.order_line.filtered(lambda line: line.product_id)
+            - self._get_reward_lines()
+        )
+        products = order_lines.mapped("product_id")
+
+        for program in applied_programs:
+
+            # First check if program still applies on delivered quantities
+
+            # To keep it simple for now only check if at least one of
+            # necessary products is delivered
+
+            valid_products = (
+                program._get_valid_products(products)
+                if program.rule_products_domain and program.rule_products_domain != "[]"
+                else products
+            )
+            qty_delivered = sum(
+                order_lines.filtered(
+                    lambda sol: sol.product_id in valid_products
+                ).mapped("qty_delivered")
+            )
+            if not qty_delivered:
+                reward_lines = self._get_program_reward_lines(program)
+                reward_lines.write({"delivered_reward_qty": 0})
+                continue
+
+            # Do we consider reward for rule_minimum_amount delivered only when
+            # this minimum amount has been delivered?
+
+            self._set_delivered_reward_qty_for_program(program)
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    qty_delivered_method = fields.Selection(
+        selection_add=[("reward_stock_move", "Reward from Stock Moves")]
+    )
+    delivered_reward_qty = fields.Float(
+        string="Prorata of reward from delivered quantities",
+        digits="Product Unit of Measure",
+    )
+
+    @api.depends("is_reward_line")
+    def _compute_qty_delivered_method(self):
+        super(SaleOrderLine, self)._compute_qty_delivered_method()
+
+        for line in self:
+            if line.is_reward_line:
+                line.qty_delivered_method = "reward_stock_move"
+
+    @api.depends("delivered_reward_qty")
+    def _compute_qty_delivered(self):
```

Comment:
Why not pivoting the whole logic from this method. I feel it would be much more simple. When delivered qties are triggered from the regular lines, this method goes into action an then you can:

- Check if there are promotions and if those promotions have a delivered qtys policy.
- One by one, decide how delivery lines are invoiced depending on the promotion type (general discount, cheapest product, specific products... and so on)

Reply:
Good idea for a future refactor